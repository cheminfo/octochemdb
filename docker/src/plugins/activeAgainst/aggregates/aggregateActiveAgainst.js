import md5 from 'md5';

import getLastDocumentImported from '../../../sync/http/utils/getLastDocumentImported.js';
import debugLibrary from '../../../utils/Debug.js';
/**
 * @description Aggregate the active against data from the activeOrNaturals collection
 * @param {*} connection
 * @output {collection} activeAgainst collection
 */
export async function aggregate(connection) {
  const debug = debugLibrary('aggregateActiveAgainst');
  try {
    // Get collections from the database
    const options = { collection: 'activeAgainst', connection };
    const progress = await connection.getProgress(options.collection);
    const progressOfSourceCollection = await connection.getProgress(
      'activesOrNaturals',
    );
    const collectionActivesOrNaturals = await connection.getCollection(
      'activesOrNaturals',
    );
    const sources = md5(progressOfSourceCollection);
    // Add logs to the collection importLogs
    const logs = await connection.getImportationLog({
      collectionName: options.collectionName,
      sources,
      startSequenceID: progress.seq,
    });
    // Get the last document imported
    const lastDocumentImported = await getLastDocumentImported(
      options.connection,
      progress,
      options.collection,
    );

    if (
      lastDocumentImported === null ||
      sources !== progress.sources ||
      progress.state !== 'updated'
    ) {
      // If the last document imported is null, or the sources are different, or the state is not updated, then we need to import the data
      const temporaryCollection = await connection.getCollection(
        `${options.collection}_tmp`,
      );
      debug.info('start Aggregation process');
      progress.state = 'aggregating';
      await connection.setProgress(progress);
      // Aggregate the data from the activesOrNaturals collection
      let dbRefs = await collectionActivesOrNaturals
        .aggregate(
          [
            { $project: { activities: '$data.activities' } },
            { $match: { activities: { $ne: [] } } },
            {
              $project: {
                _id: 0,
                activities: '$activities',
              },
            },
          ],
          {
            allowDiskUse: true, // allow aggregation to use disk if necessary
            maxTimeMS: 60 * 60 * 1000, // 1h
          },
        )
        .toArray();
      let start = Date.now();
      let count = 0;
      let total = dbRefs.length;

      dbRefs = dbRefs.filter((dbRef) => dbRef.activities !== undefined);

      for (let dbRef of dbRefs) {
        for (let activity of dbRef.activities) {
          let collectionActivity = await connection.getCollection(
            activity.collection,
          );
          let cursor = await collectionActivity.find({
            _id: activity.oid,
          });
          if (await cursor.hasNext()) {
            let entryActivity = await cursor.next();
            if (entryActivity.data?.targetTaxonomies) {
              for (let taxonomy of entryActivity.data.targetTaxonomies) {
                let entry = {
                  data: {},
                };
                if (taxonomy.superkingdom) {
                  entry.data.superkingdom = taxonomy.superkingdom;
                }
                if (taxonomy.kingdom) entry.data.kingdom = taxonomy.kingdom;
                if (taxonomy.phylum) entry.data.phylum = taxonomy.phylum;
                if (taxonomy.class) entry.data.class = taxonomy.class;
                if (taxonomy.order) entry.data.order = taxonomy.order;
                if (taxonomy.family) entry.data.family = taxonomy.family;
                if (taxonomy.genus) entry.data.genus = taxonomy.genus;
                let levels = Object.keys(entry.data);
                let id = '';
                for (let level of levels) {
                  id += `${entry.data[level]}`;
                }
                entry._id = id.replace(/\s/g, '');
                let cursor = await temporaryCollection.find({ _id: entry._id });
                if (!(await cursor.hasNext())) {
                  entry.count = 1;
                  await temporaryCollection.updateOne(
                    { _id: entry._id },
                    { $set: entry },
                    { upsert: true },
                  );
                } else {
                  await temporaryCollection.updateOne(
                    { _id: entry._id },
                    { $inc: { count: 1 } },
                    { upsert: true },
                  );
                }
                count++;
              }
            }
          }
        }
        if (Date.now() - start > Number(process.env.DEBUG_THROTTLING)) {
          await debug.info(`Aggregation progress: ${count}/${total}`);
          start = Date.now();
        }
      }

      // rename temporary collection
      await temporaryCollection.rename(options.collection, {
        dropTarget: true,
      });
      // update logs
      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'aggregated';
      await connection.updateImportationLog(logs);
      // update progress
      progress.sources = sources;
      progress.dateEnd = Date.now();
      progress.state = 'aggregated';
      await connection.setProgress(progress);
      await debug.info('Aggregation Done');
    } else {
      await debug.info(`Aggregation already up to date`);
    }
  } catch (e) {
    if (connection) {
      await debug.fatal(e.message, {
        collection: 'activeAgainst',
        connection,
        stack: e.stack,
      });
    }
  }
}
