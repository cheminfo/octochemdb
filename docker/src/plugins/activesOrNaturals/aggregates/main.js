import { cpus } from 'os';
import { Worker } from 'worker_threads';

import debugLibrary from '../../../utils/Debug.js';

export async function main(links) {
  const debug = debugLibrary('improveActivesOrNaturals Main');

  try {
    const values = Object.values(links).sort(() => Math.random() - 0.5);
    const workers = [];
    const url = new URL('worker.js', import.meta.url);

    const numWorkers = cpus().length / 2;
    const chunkSize = Math.floor(values.length / numWorkers);

    const chunks = [];
    for (let i = 0; i < numWorkers; i++) {
      const start = i * chunkSize;
      const end = i === numWorkers - 1 ? values.length : (i + 1) * chunkSize;
      const chunk = values.slice(start, end);
      chunks.push(chunk);
    }
    for (let i = 0; i < numWorkers; i++) {
      let worker = new Worker(url);
      let links = chunks[i];
      let workerID = `worker ${i + 1}`;
      worker.postMessage({ links, workerID });
      workers.push(worker);
    }

    await Promise.all(
      workers.map(
        (worker) =>
          new Promise((resolve, reject) => {
            worker.on('message', (message) => {
              debug(message);
              resolve(message);
            });
            worker.on('error', reject);
            worker.on('exit', (code) => {
              if (code !== 0) {
                reject(new Error(`Worker stopped with exit code ${code}`));
              }
            });
          }),
      ),
    );
  } catch (e) {
    debug(e);
  }
}
