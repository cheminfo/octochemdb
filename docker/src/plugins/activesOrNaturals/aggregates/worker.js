import { parentPort } from 'worker_threads';

import debugLibrary from '../../../utils/Debug.js';
import { OctoChemConnection } from '../../../utils/OctoChemConnection.js';
import getActivityKeywords from '../utils/getAcitivityKeywords.js';
import getActiveAgainstKeywords from '../utils/getActiveAgainstKeywords.js';
import getActivitiesInfo from '../utils/getActivitiesInfo.js';
import getCompoundsInfo from '../utils/getCompoundsInfo.js';
import { getMassSpectraRefForGNPs } from '../utils/getMassSpectraRefForGNPs.js';
import { getMassSpectraRefForMassBank } from '../utils/getMassSpectraRefForMassBank.js';
import { getMeshTerms } from '../utils/getMeshTerms.js';
import getTaxonomyKeywords from '../utils/getTaxonomyKeywords.js';
import getTaxonomiesInfo from '../utils/utilsTaxonomies/getTaxonomiesInfo.js';

const connection = new OctoChemConnection();
const debug = debugLibrary('WorkerProcess');
parentPort?.on('message', async (dataEntry) => {
  try {
    const { links, workerID } = dataEntry;
    debug(`Worker ${workerID} started`);
    // get worker number
    const temporaryCollection = await connection.getCollection(
      `activesOrNaturals_tmp`,
    );
    let count = 0;
    let start = Date.now();
    for (const link of links) {
      let noStereoTautomerID = link.id;
      let sources = link.sources;
      let entry = { data: { naturalProduct: false } };
      // get all documents from all collections
      let data = [];
      for (const source of sources) {
        if (
          [
            'npasses',
            'cmaups',
            'coconuts',
            'lotuses',
            'npAtlases',
            'gnps',
          ].includes(source.collection)
        ) {
          entry.data.naturalProduct = true;
        }

        const collection = await connection.getCollection(source.collection);
        let partialData = await collection.findOne({ _id: source.id });
        if (partialData) {
          partialData.collection = source.collection;
          data.push(partialData);
        }
      }
      let compoundsCollection = await connection.getCollection('compounds');
      let patentsCollection = await connection.getCollection('patents');
      let dbRefsMs = [];
      let taxons = await getTaxonomiesInfo(data, connection);
      // get unique activities from all collections for the current noStereoTautomerIDs
      let activityInfo = await getActivitiesInfo(data, connection);
      // get unique compound information from all collections for the current noStereoTautomerIDs
      entry = await getCompoundsInfo(
        entry,
        data,
        compoundsCollection,
        noStereoTautomerID,
        connection,
        patentsCollection,
      );
      let massSpectraRefsForGNPs = await getMassSpectraRefForGNPs(
        connection,
        noStereoTautomerID,
      );
      massSpectraRefsForGNPs.forEach((ref) => {
        dbRefsMs.push(ref.dbRef);
      });
      let massSpectraRefsForMassBank = await getMassSpectraRefForMassBank(
        connection,
        noStereoTautomerID,
      );
      massSpectraRefsForMassBank.forEach((ref) => {
        dbRefsMs.push(ref.dbRef);
      });
      if (entry.data.cids) {
        const uniqueMeshTerms = {};
        const uniquePmIds = {};
        let nbPubmeds = 0;
        const pubmedCollection = await connection.getCollection('pubmeds');

        for (let i = 0; i < entry.data.cids.length; i++) {
          let cid = Number(entry.data.cids[i]);
          const { meshTermsForCid, pmIds, counterPmids } = await getMeshTerms(
            cid,
            pubmedCollection,
            connection,
          );
          nbPubmeds += Number(counterPmids);
          meshTermsForCid.forEach((term) => {
            uniqueMeshTerms[term] = true;
          });
          pmIds.forEach((id) => {
            uniquePmIds[id] = true;
          });
        }

        let dbRefs = Object.keys(uniquePmIds).map((id) => ({
          $ref: 'pubmeds',
          $id: id,
        }));

        const meshTerms = Object.keys(uniqueMeshTerms);

        if (meshTerms.length > 0) {
          entry.data.kwMeshTerms = meshTerms;
        }
        if (dbRefs.length > 0) {
          entry.data.pubmeds = dbRefs;
        }
        entry.data.nbPubmeds = nbPubmeds;
      }

      if (dbRefsMs.length > 0) {
        entry.data.massSpectraRefs = dbRefsMs;
        entry.data.nbMassSpectra = dbRefsMs.length;
      }
      // if activityInfo is not empty, get unique keywords of activities and target taxonomies for the current noStereoTautomerID
      if (activityInfo.length > 0) {
        entry.data.bioActive = true;
        const keywordsActivities = getActivityKeywords(activityInfo);
        if (keywordsActivities.length > 0) {
          entry.data.kwBioassays = keywordsActivities;
        }
        const keywordsActiveAgainst = getActiveAgainstKeywords(activityInfo);

        if (keywordsActiveAgainst.length > 0) {
          entry.data.kwActiveAgainst = keywordsActiveAgainst;
        }
      }
      // if taxons is not empty, get unique keywords of taxonomies for the current noStereoTautomerID
      if (taxons.length > 0) {
        const keywordsTaxonomies = getTaxonomyKeywords(taxons);
        if (keywordsTaxonomies.length > 0) {
          entry.data.kwTaxonomies = keywordsTaxonomies;
        }
      }
      // if activityInfo is not empty, define entry.data.activities
      if (activityInfo.length > 0) {
        entry.data.activities = activityInfo;
        entry.data.nbActivities = activityInfo.length;
      }
      // if taxons is not empty, define entry.data.taxonomies
      if (taxons.length > 0) {
        entry.data.taxonomies = taxons;
        entry.data.nbTaxonomies = taxons.length;
      }

      if (entry) {
        temporaryCollection.updateOne(
          { _id: noStereoTautomerID },
          { $set: entry },
          { upsert: true },
        );

        count++;
        if (Date.now() - start > Number(process.env.DEBUG_THROTTLING)) {
          // @ts-ignore
          debug(`${workerID} aggregated ${count}`);
          start = Date.now();
        }
      }
    }
    // @ts-ignore
    parentPort.postMessage(`${workerID} aggregated ${count}`);
  } catch (e) {
    debug(e);
  }
});
