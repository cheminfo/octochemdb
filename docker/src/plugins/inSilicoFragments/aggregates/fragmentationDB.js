export const fragmentationDB = {
  ionization: {
    positive: [
      {
        Label: 'Ionization',
        rxnCode: 'eF`BLGtX!eF``fLGt#Qd Qd#!R@FL]Rgp !R@FL]vgp',
      },
      {
        Label: 'Ionization',
        rxnCode: 'eM`BN`~b@!eM``fN`~c@#Q[ Q[#!R@AM?DquRo@ !R@AM?Dqtzo@',
      },

      {
        Label: 'Ionization',
        rxnCode: 'gCh@AGj@`!gChA@Icu@_PP#qTq qTq#!RbOvw?_x@GYK| !RbOvw?_x@GWk|',
      },
      {
        Label: 'Ionization',
        rxnCode:
          'gCa@@dmPFtwhHcMCAlp!eMH`eIhOhp#Qv@ Qv#!R@AL@[@@SGSH| !R@Fp@Dqwz``',
      },
      {
        Label: 'Ionization',
        rxnCode: 'eFB@HcA}D@!eFB`HIcA}F@#QX QX#!R_vp]bgp !R_vp]vgp',
      },
      {
        Label: 'Ionization',
        rxnCode: 'eMB@HchH!eMB`HIchOh`#Q[ Q[#!R@AM?Dqtbo@ !R@AM?Dquz@`',
      },
      {
        Label: 'Ionization',
        rxnCode: 'eMHAIXMicLLG~r!eFH`fJGtP#QX QX#!R@AL@[AtbO@ !R@AL]nkp',
      },
      {
        Label: 'Ionization',
        rxnCode: 'eMHBN``!eMH`fN`~b@#Qg Qg#!R@FL?XqtJ_@ !R@FL?XqwZ_@',
      },
      {
        Label: 'Ionization',
        rxnCode: 'eFB@HcA}D@!eFB`HIcA}F@#QX QX#!R_vp]bgp !R_vp]vgp',
      },
      {
        Label: 'Ionization',
        rxnCode: 'eFHBLGtP!eFH`fLGtX#QX QX#!R@AL]Pmp !R@AL]^gp',
      },
      {
        Label: 'Ionization',
        rxnCode: 'eM`BN`~b@!eM``fN`~bOza@#Qg Qg#!R@Fq?[AuJ?@ !R@FM?Dqvz_@',
      },
    ],
  },
  positive: {},
};
