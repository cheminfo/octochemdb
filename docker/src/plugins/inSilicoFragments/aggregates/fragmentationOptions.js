export let fragmentationOptions = {
  maxDepth: 3,
  limitReactions: 500,
  minIonizations: 1,
  maxIonizations: 1,
  minReactions: 0,
  maxReactions: 2,
};
