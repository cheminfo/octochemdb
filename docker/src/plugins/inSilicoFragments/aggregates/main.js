import { cpus } from 'os';
import { Worker } from 'worker_threads';

import debugLibrary from '../../../utils/Debug.js';
import { OctoChemConnection } from '../../../utils/OctoChemConnection.js';

const connection = new OctoChemConnection();
export async function main(links) {
  const debug = debugLibrary('improveInSilicoFragments Main');

  try {
    let total = Object.keys(links).length;

    let values = Object.values(links).sort(() => Math.random() - 0.5);

    const workers = [];
    const url = new URL('worker.js', import.meta.url);

    const numWorkers = cpus().length / 2;
    const chunkSize = Math.floor(values.length / numWorkers);
    debug.trace(`Starting ${numWorkers} workers`);
    debug.trace(`Chunk size: ${chunkSize}`);
    debug.trace(`Total: ${total}`);
    for (let i = 0; i < numWorkers; i++) {
      const start = i * chunkSize;
      const end = i === numWorkers - 1 ? values.length : (i + 1) * chunkSize;
      const chunk = values.slice(start, end);
      let worker = new Worker(url);
      worker.postMessage({ links: chunk, workerID: i });
      workers.push(worker);
    }

    const counts = new Uint32Array(numWorkers);
    let lastLogDate = Date.now();
    await Promise.all(
      workers.map(
        (worker) =>
          new Promise((resolve, reject) => {
            worker.on('message', (message) => {
              counts[message.workerID] = message.currentCount;
              if (
                Date.now() - lastLogDate >
                  Number(process.env.DEBUG_THROTTLING) &&
                message.status === 'running'
              ) {
                let current = counts.reduce(
                  (previous, current) => previous + current,
                  0,
                );
                lastLogDate = Date.now();
                debug.trace(`Fragmented: ${current} / ${total} Molecules`);
              }
              if (message.status === 'done') {
                resolve(message);
              }
            });
            worker.on('error', (err) => {
              debug.fatal(err.message, {
                collection: 'inSilicoFragments',
                connection,
                stack: err.stack,
              });
            });
            worker.on('exit', (code) => {
              debug.trace(code);
              if (code !== 0) {
                reject(new Error(`Worker stopped with exit code ${code}`));
              }
            });
          }),
      ),
    );
  } catch (e) {
    if (connection) {
      await debug.fatal(e.message, {
        collection: 'inSilicoFragments',
        connection,
        stack: e.stack,
      });
    }
  }
}
