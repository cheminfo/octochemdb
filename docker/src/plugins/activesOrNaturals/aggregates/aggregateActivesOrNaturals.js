import md5 from 'md5';

import debugLibrary from '../../../utils/Debug.js';
import getCollectionsLinks from '../utils/getCollectionsLinks.js';

import { main } from './main.js';
/**
 * @description Aggregate all synchronized collections into one collection of Bioactivities or/and Natural Products
 * @param {*} connection MongoDB connection
 * @return {Promise} Returns ActiveOrNaturals collection
 */
export async function aggregate(connection) {
  const collectionNames = [
    'lotuses',
    'npasses',
    'npAtlases',
    'cmaups',
    'coconuts',
    'bioassays',
  ];
  const debug = debugLibrary('aggregateActivesOrNaturals');
  try {
    const options = { collection: 'activesOrNaturals', connection };
    // Get progress,logs, target, lastDocument and links of the collection
    const progress = await connection.getProgress(options.collection);
    const targetCollection = await connection.getCollection(options.collection);

    let { links, collectionSources } = await getCollectionsLinks(
      connection,
      collectionNames,
    );
    const pubmedProgress = await connection.getProgress('pubmeds');
    collectionSources.push(pubmedProgress.sources);
    const sources = md5(collectionSources);
    const logs = await connection.getImportationLog({
      collectionName: options.collectionName,
      sources,
      startSequenceID: progress.seq,
    });

    if (sources !== progress.sources || progress.state !== 'aggregated') {
      // if lastDocumentImported is null or sources are different from the progress, start aggregation process
      const temporaryCollection = await connection.getCollection(
        `${options.collection}_tmp`,
      );

      // debug unique numbers of noStereoTautomerIDs
      debug.trace(
        `Unique numbers of noStereoTautomerIDs: ${Object.keys(links).length}`,
      );
      debug.info('start Aggregation process');
      // set progress to aggregating

      progress.state = 'aggregating';
      await connection.setProgress(progress);
      // parse all noStereoTautomerIDs and get their info
      await main(links);
      // rename temporary collection to activesOrNaturals
      await temporaryCollection.rename(options.collection, {
        dropTarget: true,
      });
      // set logs to aggregated
      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'aggregated';
      await connection.updateImportationLog(logs);
      // set progress to aggregated
      progress.sources = sources;
      progress.dateEnd = Date.now();
      progress.state = 'aggregated';
      await connection.setProgress(progress);

      // Create Indexes
      await targetCollection.createIndex({ 'data.em': 1 });
      await targetCollection.createIndex({ 'data.mf': 1 });
      await targetCollection.createIndex({ 'data.kwBioassays': 1 });
      await targetCollection.createIndex({ 'data.meshTerms': 1 });
      await targetCollection.createIndex({ 'data.kwTaxonomies': 1 });
      await targetCollection.createIndex({ 'data.nbActivities': 1 });
      await targetCollection.createIndex({ 'data.pubmeds': 1 });
      await targetCollection.createIndex({ 'data.nbPubmeds': 1 });
      await targetCollection.createIndex({ 'data.patents': 1 });
      await targetCollection.createIndex({ 'data.nbPatents': 1 });
      await targetCollection.createIndex({ 'data.nbTaxonomies': 1 });
      await targetCollection.createIndex({ 'data.nbMassSpectra': 1 });
      await targetCollection.createIndex({ 'data.molecules': 1 });
      await targetCollection.createIndex({ 'data.nbMolecules': 1 });

      debug.info('Aggregation Done');
    } else {
      debug.info(`Aggregation already up to date`);
    }
  } catch (e) {
    if (connection) {
      await debug.fatal(e.message, {
        collection: 'activesOrNaturals',
        connection,
        stack: e.stack,
      });
    }
  }
}
