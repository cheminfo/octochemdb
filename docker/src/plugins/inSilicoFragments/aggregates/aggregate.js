import md5 from 'md5';

import getLastDocumentImported from '../../../sync/http/utils/getLastDocumentImported.js';
import debugLibrary from '../../../utils/Debug.js';
import getCollectionsLinks from '../utils/getCollectionsLinks.js';

import { main } from './main.js';

export async function aggregate(connection) {
  const debug = debugLibrary('inSilicoFragments');
  try {
    // Get collections from the database
    const options = { collection: 'inSilicoFragments', connection };
    const progress = await connection.getProgress(options.collection);
    const progressOfSourceCollection = await connection.getProgress(
      'activesOrNaturals',
    );

    const sources = md5(progressOfSourceCollection);
    // Add logs to the collection importLogs
    const logs = await connection.getImportationLog({
      collectionName: options.collectionName,
      sources,
      startSequenceID: progress.seq,
    });
    // Get the last document imported
    const lastDocumentImported = await getLastDocumentImported(
      options.connection,
      progress,
      options.collection,
    );

    if (
      lastDocumentImported === null ||
      sources !== progress.sources ||
      progress.state !== 'updated'
    ) {
      const temporaryCollection = await connection.getCollection(
        `${options.collection}_tmp`,
      );
      debug.info('start fragmentation process');
      progress.state = 'aggregating';
      await connection.setProgress(progress);

      let links = await getCollectionsLinks(connection);

      await main(links);
      await temporaryCollection.createIndex({ 'data.masses.positive': 1 });
      await temporaryCollection.createIndex({ 'data.ocl.idCode': 1 });
      // rename temporary collection
      await temporaryCollection.rename(options.collection, {
        dropTarget: true,
      });
      // update logs
      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'aggregated';
      await connection.updateImportationLog(logs);
      // update progress
      progress.sources = sources;
      progress.dateEnd = Date.now();
      progress.state = 'aggregated';
      await connection.setProgress(progress);
      await debug.info('Aggregation Done');
    } else {
      await debug.info(`Aggregation already up to date`);
    }
  } catch (e) {
    if (connection) {
      await debug.fatal(e.message, {
        collection: 'inSilicoFragments',
        connection,
        stack: e.stack,
      });
    }
  }
}
