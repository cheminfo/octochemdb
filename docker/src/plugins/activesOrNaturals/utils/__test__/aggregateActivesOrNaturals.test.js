import { test, expect } from 'vitest';

import { OctoChemConnection } from '../../../../utils/OctoChemConnection.js';
import { aggregate } from '../../aggregates/aggregateActivesOrNaturals';

test(
  'Aggregation ActivesOrNaturals',
  async () => {
    const connection = new OctoChemConnection();
    const lotusesCollection = await connection.getCollection('lotuses');
    const npassesCollection = await connection.getCollection('npasses');
    const npAtlasesCollection = await connection.getCollection('npAtlases');
    const cmaupsCollection = await connection.getCollection('cmaups');
    const coconutsCollection = await connection.getCollection('coconuts');
    const bioassaysCollection = await connection.getCollection('bioassays');
    const gnpsCollection = await connection.getCollection('gnps');
    const pubmedsCollection = await connection.getCollection('pubmeds');
    const patentsCollection = await connection.getCollection('patents');
    const compoundPatentsCollection = await connection.getCollection(
      'compoundPatents',
    );
    while (true) {
      if (
        (await lotusesCollection.countDocuments()) === 20 &&
        (await npassesCollection.countDocuments()) === 12 &&
        (await npAtlasesCollection.countDocuments()) === 3 &&
        (await cmaupsCollection.countDocuments()) === 19 &&
        (await coconutsCollection.countDocuments()) === 9 &&
        (await bioassaysCollection.countDocuments()) === 20 &&
        (await gnpsCollection.countDocuments()) === 2 &&
        (await pubmedsCollection.countDocuments()) === 7 &&
        (await patentsCollection.countDocuments()) === 255 &&
        (await compoundPatentsCollection.countDocuments()) === 4502
      ) {
        break;
      }
    }

    await aggregate(connection);
    const collection = await connection.getCollection('activesOrNaturals');
    const collectionEntry = await collection
      .find({
        _id: 'ekTpA@@@LAEMGLn\\dTTRbRfLbteRrRTfbqbtRthdRjZFFfNnAQjjjjjjjfjjjjjijjh@@',
      })
      .limit(1);

    let result = await collectionEntry.next();
    if (result?._seq) {
      delete result._seq;
    }
    if (result) {
      for (const molecule of result.data.molecules) {
        delete molecule.ocl.coordinates;
      }
      delete result.data.noStereoOCL.coordinates;
    }
    expect(result).toMatchSnapshot();
    await connection.close();
  },
  { timeout: 50000 },
);
