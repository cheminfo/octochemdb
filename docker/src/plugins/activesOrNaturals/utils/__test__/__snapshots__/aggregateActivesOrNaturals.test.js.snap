// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Aggregation ActivesOrNaturals 1`] = `
{
  "_id": "ekTpA@@@LAEMGLn\\\\dTTRbRfLbteRrRTfbqbtRthdRjZFFfNnAQjjjjjjjfjjjjjijjh@@",
  "data": {
    "activities": [
      {
        "dbRef": {
          "$id": "LTS0257199",
          "$ref": "lotuses",
        },
      },
      {
        "dbRef": {
          "$id": "59478_1",
          "$ref": "bioassays",
        },
      },
      {
        "dbRef": {
          "$id": "61929_1",
          "$ref": "bioassays",
        },
      },
      {
        "dbRef": {
          "$id": "67115_1",
          "$ref": "bioassays",
        },
      },
      {
        "dbRef": {
          "$id": "68553_1",
          "$ref": "bioassays",
        },
      },
      {
        "dbRef": {
          "$id": "76598_1",
          "$ref": "bioassays",
        },
      },
      {
        "dbRef": {
          "$id": "77111_1",
          "$ref": "bioassays",
        },
      },
      {
        "dbRef": {
          "$id": "77250_1",
          "$ref": "bioassays",
        },
      },
      {
        "dbRef": {
          "$id": "80565_1",
          "$ref": "bioassays",
        },
      },
      {
        "dbRef": {
          "$id": "81146_1",
          "$ref": "bioassays",
        },
      },
      {
        "dbRef": {
          "$id": "84145_1",
          "$ref": "bioassays",
        },
      },
      {
        "dbRef": {
          "$id": "88579_1",
          "$ref": "bioassays",
        },
      },
      {
        "dbRef": {
          "$id": "89216_1",
          "$ref": "bioassays",
        },
      },
      {
        "dbRef": {
          "$id": "90481_1",
          "$ref": "bioassays",
        },
      },
      {
        "dbRef": {
          "$id": "95282_1",
          "$ref": "bioassays",
        },
      },
      {
        "dbRef": {
          "$id": "96028_1",
          "$ref": "bioassays",
        },
      },
      {
        "dbRef": {
          "$id": "96679_1",
          "$ref": "bioassays",
        },
      },
      {
        "dbRef": {
          "$id": "5351335_1",
          "$ref": "bioassays",
        },
      },
      {
        "dbRef": {
          "$id": "5351409_1",
          "$ref": "bioassays",
        },
      },
      {
        "dbRef": {
          "$id": "5351514_1",
          "$ref": "bioassays",
        },
      },
      {
        "dbRef": {
          "$id": "5351641_1",
          "$ref": "bioassays",
        },
      },
    ],
    "bioactive": true,
    "charge": 0,
    "cids": [
      {
        "$id": 59478,
        "$ref": "compounds",
      },
    ],
    "em": 632.42881889036,
    "kwActiveAgainst": [
      "archaea",
      "borrarchaeaceae",
      "borrarchaeales",
      "borrarchaeia",
      "borrarchaeota",
      "borrarchaeum",
      "candidatus",
    ],
    "kwBioassays": [
      "eukaryotic",
      "inhibitors",
      "initiation",
      "molecule",
      "small",
      "translation",
      "uhts",
    ],
    "kwTaxonomies": [
      "viridiplantae",
      "streptophyta",
      "magnoliopsida",
      "cucurbitaceae",
      "momordica",
      "charantia",
    ],
    "mf": "C37H60O8",
    "naturalProduct": true,
    "nbActivities": 21,
    "nbMassSpectra": 0,
    "nbPatents": 3,
    "nbPubmeds": 0,
    "nbTaxonomies": 1,
    "noStereoOCL": {
      "idCode": "ekTpA@@@LAEMGLn\\\\dTTRbRfLbteRrRTfbqbtRthdRjZFFfNnAQjjjjjjjfjjjjjijjh@@",
    },
    "patents": [
      {
        "$id": "EP-2078065-A2",
        "$ref": "patents",
      },
    ],
    "taxonomies": [
      {
        "class": "Magnoliopsida",
        "dbRef": {
          "$id": "LTS0257199",
          "$ref": "lotuses",
        },
        "family": "Cucurbitaceae",
        "genus": "Momordica",
        "kingdom": "Viridiplantae",
        "phylum": "Streptophyta",
        "species": "Momordica charantia",
      },
    ],
    "unsaturation": 8,
  },
}
`;
